src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -30,6 +34,170 @@
+          self._chroot.write(provider.get_resource_string(source_name, rel_path),
+                             os.path.join(BOOTSTRAP_DIR, source_name, rel_path),
+                             'bootstrap')
+
+    # Setup a re-director package to support the legacy pex runtime `_pex` APIs through a
+    # VendorImporter.
+    self._chroot.touch(os.path.join(BOOTSTRAP_DIR, LEGACY_BOOSTRAP_PKG, '__init__.py'), 'bootstrap')
+
+
+# This bootstrapping technique for lazily loading requirements comes from the pex project test case:
+# https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+_IPEX_PREAMBLE = """\
+import json
+import os
+import sys
+
+from pex import resolver, vendor
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import BOOTSTRAP_DIR, PEXBuilder
+from pex.pex_info import PexInfo
+from pex.platforms import Platform
+from pex.third_party.pkg_resources import DefaultProvider, ZipProvider, get_provider
+from pex.util import CacheHelper
+from pex.variables import ENV
+
+# NB: Keep this in sync between pex_build_util.py and the _IPEX_PREAMBLE!
+class IPEXBuilder(PEXBuilder):
+  "Custom PEXBuilder subclass to expose the 'wheel' and 'vendor' packages."
+
+  def _prepare_bootstrap(self):
+    from pex import vendor
+    from pex.pex_builder import BOOTSTRAP_DIR, LEGACY_BOOSTRAP_PKG
+    from pex.third_party.pkg_resources import DefaultProvider, ZipProvider, get_provider
+
+    try:
+      vendor.vendor_runtime(chroot=self._chroot,
+                            dest_basedir=BOOTSTRAP_DIR,
+                            label='bootstrap',
+                            # NB: We use wheel here in the builder, but that's only at build-time.
+                            root_module_names=['pkg_resources', 'wheel'])
+    except ValueError:
+      pass
+
+    source_name = 'pex'
+    provider = get_provider(source_name)
+    if not isinstance(provider, DefaultProvider):
+      mod = __import__(source_name, fromlist=['ignore'])
+      provider = ZipProvider(mod)
+    for package in ('', 'third_party', 'vendor'):
+      for fn in provider.resource_listdir(package):
+        if fn.endswith('.py'):
+          rel_path = os.path.join(package, fn)
+          self._chroot.write(provider.get_resource_string(source_name, rel_path),
+                             os.path.join(BOOTSTRAP_DIR, source_name, rel_path),
+                             'bootstrap')
+
+    # Setup a re-director package to support the legacy pex runtime `_pex` APIs through a
+    # VendorImporter.
+    self._chroot.touch(os.path.join(BOOTSTRAP_DIR, LEGACY_BOOSTRAP_PKG, '__init__.py'), 'bootstrap')
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+assert ext != '.pex', "Could not hydrate ipex: {}.pex must not be named with the .pex extension.".format(filename_base)
```

Comment:
Do you want to always preserve this check? If so, use an exception. assertions can be turned off in Python when running in production mode.

Reply:
Turned into `ValueError`s!